dpnp/dpnp_iface_histograms.py

```diff
@@ -0,0 +1,336 @@
+    """Computes the bins used internally by `histogram`."""
+
+    # parse the overloaded bins argument
+    n_equal_bins = None
+    bin_edges = None
+    sycl_queue = a.sycl_queue
+
+    if isinstance(bins, str):
+        raise NotImplementedError("only integer and array bins are implemented")
+
+    if numpy.ndim(bins) == 0:
+        try:
+            n_equal_bins = operator.index(bins)
+        except TypeError as e:
+            raise TypeError("`bins` must be an integer or an array") from e
+        if n_equal_bins < 1:
+            raise ValueError("`bins` must be positive, when an integer")
+
+        first_edge, last_edge = _get_outer_edges(a, range)
+
+    elif numpy.ndim(bins) == 1:
+        if dpnp.is_supported_array_type(bins):
+            if dpu.get_execution_queue([a.sycl_queue, bins.sycl_queue]) is None:
+                raise ValueError(
+                    "a and bins must be allocated on the same SYCL queue"
+                )
+
+            bin_edges = bins
+        else:
+            bin_edges = dpnp.asarray(
+                bins, sycl_queue=sycl_queue, usm_type=usm_type
+            )
+
+        if dpnp.any(bin_edges[:-1] > bin_edges[1:]):
+            raise ValueError(
+                "`bins` must increase monotonically, when an array"
+            )
+
+    else:
+        raise ValueError("`bins` must be 1d, when an array")
+
+    if n_equal_bins is not None:
+        # numpy's gh-10322 means that type resolution rules are dependent on
+        # array shapes. To avoid this causing problems, we pick a type now and
+        # stick with it throughout.
+        bin_type = dpnp.result_type(first_edge, last_edge, a)
+        if dpnp.issubdtype(bin_type, dpnp.integer):
+            bin_type = dpnp.result_type(
+                bin_type, dpnp.default_float_type(sycl_queue=sycl_queue), a
+            )
+
+        # bin edges must be computed
+        bin_edges = dpnp.linspace(
+            first_edge,
+            last_edge,
+            n_equal_bins + 1,
+            endpoint=True,
+            dtype=bin_type,
+            sycl_queue=sycl_queue,
+            usm_type=usm_type,
+        )
+        return bin_edges, (first_edge, last_edge, n_equal_bins)
+    return bin_edges, None
+
+
+def _search_sorted_inclusive(a, v):
+    """
+    Like :obj:`dpnp.searchsorted`, but where the last item in `v` is placed
+    on the right.
+    In the context of a histogram, this makes the last bin edge inclusive
+
+    """
+
+    return dpnp.concatenate(
+        (a.searchsorted(v[:-1], "left"), a.searchsorted(v[-1:], "right"))
+    )
+
+
+def histogram(a, bins=10, range=None, density=None, weights=None):
+    """
+    Compute the histogram of a dataset.
+
+    For full documentation refer to :obj:`numpy.histogram`.
+
+    Parameters
+    ----------
+    a : {dpnp.ndarray, usm_ndarray}
+        Input data. The histogram is computed over the flattened array.
+    bins : {int, dpnp.ndarray, usm_ndarray, sequence of scalars}, optional
```

Comment:
I've placed also `TODO` comment in the code

Reply:
That is fine. I just mentioned it because we say 
> If `bins` is a string, it defines the method used to calculate the optimal bin width, as defined by :obj:`dpnp.histogram_bin_edges`.

in the explanation.

Reply:
I will remove that from the description.